### DISCUSSION TITLE
Any way to save the trained volume as a VDB file?

### DISCUSSION THREAD

**Discussion by hradec** [NVlabs/instant-ngp#414] [Q&A]
Before anything, this is incredible!!! Congratulations for your hard work, and thank you so much for making it available here on github!! 

I'm wondering if it would be possible to export the trained volume as a VDB file, to import on other software, like Maya/Houdini, but most important, to render on offline rendering software, like renderman, cycles and arnold. 

I'm actually very interested in test the potential of using this VDB as a volume light emitter to use for VFX rendering integration. Basically, this could be the perfect substitute for HDRI dome lighting, with more correct lighting results for complex live action scenarios.

**Comment by Tom94** [comment]
Hi, there. At the moment, this isn't possible with a simple command, but it should be reasonably easy to do programmatically. You could add a function to `python_api.cu`, which evaluates the density of the learned model on a dense grid (similar to the first step of the marching cubes function).

Then, on the Python side, you could simply populate a vdb grid with the resulting values and save it out. (See https://www.openvdb.org/documentation/doxygen/python.html )

Cheers!

**Reply by hradec** [reply]
> Hi, there. At the moment, this isn't possible with a simple command, but it should be reasonably easy to do programmatically. You could add a function to `python_api.cu`, which evaluates the density of the learned model on a dense grid (similar to the first step of the marching cubes function).
> 
> Then, on the Python side, you could simply populate a vdb grid with the resulting values and save it out. (See https://www.openvdb.org/documentation/doxygen/python.html )

That sounds perfect! I'll have a look at it... thanks lots matte! 

cheers!

**Reply by paul-ferragut** [reply]
It would be great to save the trained volume as a VDB file... It would open new creative opportunities to be able to use nerf vdb with Unreal/Unity/Houdini... I dream this feature will be added soon 🤞🤞

Dear @hradec thank you for creating the VDB-export fork, did you managed to successfully export VDB?
I tried it but have an issue with pyopenvdb. I have pyopenvdb installed with `pip install pyopenvdb-3.8`
`Traceback (most recent call last):
  File "scripts\run.py", line 307, in <module>
    import pyopenvdb as vdb
ModuleNotFoundError: No module named 'pyopenvdb'`
It might be because I am using Windows and I am looking for solutions

**Reply by carlosedubarreto** [reply]
There is a fork for vdb export?
Where can I find it?
@paul-ferragut , can you share the link please?

**Reply by paul-ferragut** [reply]
@carlosedubarreto https://github.com/hradec/instant-ngp/tree/vdb-export

**Reply by carlosedubarreto** [reply]
Thanks a lot!!! @paul-ferragut

**Reply by carlosedubarreto** [reply]
@paul-ferragut 
from what I saw, there is no packing for windows of pyopenvdb.
So, whet you will probably need to do is to compile it yourself.
But there is a (not too detailed) setp by step explaining how to build on windows.

here is the link
https://github.com/AcademySoftwareFoundation/openvdb#windows

I'm trying to build it myself, I thing that with this build it will also come with the python bindings, the same way as intant-ngp came.

**Reply by hradec** [reply]
> Dear @hradec thank you for creating the VDB-export fork, did you managed to successfully export VDB? I tried it but have an issue with pyopenvdb. I have pyopenvdb installed with `pip install pyopenvdb-3.8` `Traceback (most recent call last): File "scripts\run.py", line 307, in <module> import pyopenvdb as vdb ModuleNotFoundError: No module named 'pyopenvdb'` It might be because I am using Windows and I am looking for solutions

Yes I did! Sorry it took me so long to reply... life is hectic and after I got the VDB exporting working, I had to drop it to focus of something else. 

I just submitted my latest changes to the vdb-export branch. It's still outdated with the instantNGP code from jun/2022, but it's all there now. 

The VDB exporting is done correctly with colors too now. 

There's a build.sh that builds all the necessary dependencies and instantNGP itself, including openvdb. 
The build.sh will try to download everything it needs and build it locally, including python, openvdb and pyopenvdb. 

If everything builds ok, you should be able to run any of the scripts by using the run.sh script. That script configures all env vars so it runs with the programs and libraries built by build.sh, including python!

You may need to do some adjustments to the env.sh script before running build.sh, since I wrote it to use my own version of GCC that we have setup in our pipeline... so you may want to get rid of that. The lines are
```
# now we setup pipevfx gcc to build with
export GCC_VERSION=9.3.1
export GCC_TARGET_FOLDER="$PIPE/gcc/$GCC_VERSION/"
export _CXXFLAGS=" -isystem$GCC_TARGET_FOLDER/lib/gcc/x86_64-pc-linux-gnu/$GCC_VERSION/include-fixed/\
                             -I$GCC_TARGET_FOLDER/include/c++/$GCC_VERSION/\
                             -isystem$GCC_TARGET_FOLDER/include/c++/$GCC_VERSION/\
                             -isystem$GCC_TARGET_FOLDER/include/c++/$GCC_VERSION/x86_64-pc-linux-gnu/\
                             -isystem$GCC_TARGET_FOLDER/lib/gcc/x86_64-pc-linux-gnu/$GCC_VERSION/include/\
                             -isystem$GCC_TARGET_FOLDER/lib/gcc/x86_64-pc-linux-gnu/$GCC_VERSION/include/c++\
                             -isystem$GCC_TARGET_FOLDER/lib/gcc/x86_64-pc-linux-gnu/$GCC_VERSION/include/c++/x86_64-pc-linux-gnu/\
                             -isystem/usr/include "


export PATH=$GCC_TARGET_FOLDER/bin:$PATH
```

After running build.sh (and if everything builds OK), you can run make.sh to test it. The make.sh is the script I used to re-build and test during development. So running make.sh will run make again in the build folder, and run run.sh to execute instantNGP with the fox example, and export the fox nerf as a VDB! 

I think that's pretty much it. 

Again, a thousand apologies for the long time to reply... and let me know if you need any help! I'll try to be more careful when going through my github notifications so I don't miss any future replies! 

PS: this is all for LINUX... If you need on windows, I can't help you matte. :(

**Reply by carlosedubarreto** [reply]
Wow, great news.
I suffered a lot to build your solution on windows. and sadly the final result was not good enough (at least for me 😊)
Do you know if the result you've got from the version that you updated with color can have a more detailed output?

I would not like to annoy you too much, but do you have an vdb export example that I could try?
So I dont suffer again to built your new code on windows and and up having a similar result I had last time 😊

And thanks a lot for the information

**Comment by carlosedubarreto** [comment]
@paul-ferragut if you like, you can get the compilation I did with python 3.10

I added the files in my page
https://www.patreon.com/posts/72933009

**Comment by carlosedubarreto** [comment]
@Tom94 , I saw that @hradec made a great job by making the export of the nerf data to volume, but he made an export only of the density, as you can see in this section
![image](https://user-images.githubusercontent.com/4061130/194430408-20267f4a-c29f-4f65-926a-8b4cdff05c5f.png)

I know almost nothing bout this, but I'm willing to try to improve to get the colors. Do you have an idea of what I could do to export the colors?
Where I could research it?

Whit what he did, I could get it inside of blender but its dificult to use like this
![image](https://user-images.githubusercontent.com/4061130/194430681-b493dc15-1974-402f-aee7-2362182c5e15.png)

from the picture I sent you can see the head of the fox that comes as example
and if it server for something below is the vdb result

[out.zip](https://github.com/NVlabs/instant-ngp/files/9729031/out.zip)

Thanks a lot.

**Reply by carlosedubarreto** [reply]
Found how to export the colors, here is an example in blender

![image](https://user-images.githubusercontent.com/4061130/194509608-db7a8ab1-a321-406c-aa82-2da0bfed5244.png)

**Reply by adkAurora** [reply]
i have the same problem,   what I could do to export the colors? do i need to change the export code? how can i get colorful results like you?

**Reply by carlosedubarreto** [reply]
@liyujiejiejie , to be honest I gave up on that, I wasnt able to mke something good, but found another person that did, and inside blender.

Take a look on his approach, https://github.com/JamesPerlman/blender_nerf_tools

**Reply by adkAurora** [reply]
i just want to get the results above like you, could you please tell me

**Comment by hradec** [comment]
Sorry the long delay to reply, guys. 

I have just submitted all my changes to my vdb fork, with color exporting too. 
https://github.com/NVlabs/instant-ngp/discussions/414#discussioncomment-6053864

All the build is done in Linux, so I can't help if you're in windows. :(

I need to find time now to update the code on my fork with the latest from NVLabs to submit a pull request to the main master branch.

**Reply by Nomad1** [reply]
Any updates on this? I see the vdb-export branch in your github but it was never pulled to the main tree if I got it right. Will give it a try shortly. Nevertheless, it's a great feature if working as expected but it needs a bit of documentation and samples.
Thanks!
